keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,402 @@
+starts, all old data is discarded, so an inline ephemeral volume is a
+suitable abstraction for declaring the need for a volume that is
+backed by PMEM and provided by
+[PMEM-CSI](https://github.com/intel/pmem-csi). But PMEM is a resource
+that is local to a node and thus the scheduler has to be aware whether
+enough of it is available on a node before assigning a pod to it.
+
+
+### Caching remaining capacity via the API server
+
+CSI defines the GetCapacity RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. A new API type which represents capacity
+is introduced to solve this (details below).
+
+The CSI driver deployment is responsible for creating and updating
+objects of that type. The external-provisioner gets extended to handle
+this, so CSI drivers only need to implement the GetCapacity call. Each
+capacity value has a lifetime chosen by the creator. Removal of
+expired values is left to the Kubernetes scheduler. This avoids the
+issue that two different processes both try the same operation, which
+would be redundant work. Making the Kubernetes scheduler responsible
+for this has the advantage that it also works when the driver crashes
+or gets de-installed.
+
+The Kubernetes scheduler monitors those objects and uses that
+information when it comes to making scheduling decisions. If the
+scheduler has no current capacity information, it proceeds without it,
+so nothing changes for CSI driver deployments that do not support
+capacity tracking.
+
+This is a problem for deployments that do support it, but haven’t been
+able to provide the necessary information yet. TODO: extend
+CSIDriverInfo with a field that tells the scheduler to retry later
+when information is missing? Avoid the situation by (Somehow? How?!)
+always providing the information before registering the driver -
+probably doesn’t work for PVCs with late binding because the pod gets
+scheduled even if the driver isn’t installed (which also is a problem
+for the CSIDriverInfo approach…).
+
+### API
+
+All Kubernetes API changes are under the `CSIStorageCapacity` feature
+gate.
+
+#### Storage capacity
+
+The prefix “CSI” is used to avoid confusion with other types that have
+a similar name and because some aspects are specific to CSI.
+
+```
+// CSIDriverCapacity represents information about storage capacity that is available
+// via a certain CSI driver. Like CSIDriverInfo, the name of the CSIDriverCapacity
+// objects is the same as the driver name. Unlike CSIDriverInfo, these objects are
+// created dynamically and get updated regularly when there are changes.
+type CSIDriverCapacity struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+   // Capacity contains the current capacity available via the driver for each
+   // storage class that uses the driver. The empty class name is
+   // valid and is used for ephemeral inline volumes.
+   Capacity: map[string]CSITopologyCapacity
```

Comment:
why you prefer putting all the information together?

Reply:
You mean why I propose to have one `CSIDriverCapacity` object per CSI driver with multiple `CSITopologyCapacity` inside?

The main motivation was that it avoids the naming problem for those objects. They can use the same name as the CSI driver because that already has to be unique. It will lead to nice output for `go get csidrivercapacity` compared to some other approach where the name is derived from CSI driver name plus storage class name.

It obviously has implications for the number of objects and updates that the API server has to handle. Here I am not sure what is better: many smaller objects that can be updated independently or fewer big objects which get updated more often.

Reply:
I'm experimenting with enabling patching for that big object. That should address most of the concerns that I had myself about this approach:
- we can have multiple writers which do not need the latest copy of the entire object when changing the subset of the data that they are responsible for
- each individual update is small (just a PATCH operation with a minimal merge patch).